pmaudit

```diff
@@ -415,50 +415,59 @@ def main():
         '-W', '--watch', action='append', default=[os.curdir],
         metavar='DIR',
         help="audit activity within DIRs (default=%(default)s)")
+    parser.add_argument( # default='%s.json' % PROG
+        '-o', '--save', metavar='FILE',
+        help="save audit data in JSON format to FILE")
+    parser.add_argument(
+        '-d', '--depsfile', metavar='FILE',
+        help="save audit data in makefile format to FILE")
 
     if '--' in sys.argv or '-c' in sys.argv or '--cmd' in sys.argv:
         if '--' in sys.argv:
-            parser.add_argument(
-                '-o', '--save', default='%s.json' % PROG,
-                metavar='FILE',
-                help="save audit data to FILE (default=%(default)s)")
             opts, cmd = parser.parse_known_args()
             cfglog(opts.verbosity)
             if '--' in cmd:
                 cmd.remove('--')
             assert not (opts.cmd or opts.errexit)
             sys.stderr.write('+ %s\n' % ' '.join(cmd))
         else:
-            parser.add_argument(
-                '-o', '--save', required=True,
-                metavar='FILE',
-                help="save audit data to FILE")
             opts = parser.parse_args()
-            cmd = [os.getenv('SHELL', '/bin/sh')]
             cfglog(opts.verbosity)
+            cmd = [os.getenv('SHELL', '/bin/sh')]
             if opts.errexit:
                 cmd.append('-e')
             if opts.verbosity:
                 cmd.append('-x')
             cmd += ['-c', opts.cmd]
+        if not opts.save and not opts.depsfile: # Implement default save.
+            opts.save = '%s.json' % PROG
         wdirs = []
         for word in opts.watch:
             wdirs.extend(word.split(','))
-        audit = PMAudit(wdirs, exclude=(os.path.basename(opts.save),))
+        exclude_set = set()
+        if opts.save:
+            exclude_set |= {opts.save}
+        if opts.depsfile:
+            exclude_set |= {opts.depsfile}
+        audit = PMAudit(wdirs, exclude=exclude_set)
         audit.start(flush_host=opts.flush_host, keep_going=opts.keep_going)
         rc = subprocess.call(cmd)
         adb = audit.finish(cmd=opts.cmd or ' '.join(cmd))
-        if opts.cmd:
+        if opts.depsfile:
             prqs = adb[DB][PREREQS]
+            # TODO: Perhaps we should *always* create depsfile, even if empty.
             if prqs:
-                with open(opts.save, 'w') as f:
-                    f.write(os.path.splitext(opts.save)[0] + ': \\\n')
+                depsdir = os.path.dirname(opts.depsfile)
+                if depsdir and not os.path.exists(depsdir):
```

Comment:
Always a little bit of a race between testing and creating ... directory creation is one of the trickiest things to get right. Here I think the best practice is something like:

try:
    os.makedirs(depsdir)
except OSError:
    if not os.path.isdir(depsdir):
        raise

Reply:
Ok!  I think I'll make a function like mkdir_p_or_die() because the making directories happens more than once.

Reply:
Thanks.

Reply:
Wrapping makedirs with an exception handler is the way to go, but it's best to also use makedir's "exist_ok=True" parameter (added in Python 3.2, so I think that's okay to use).  I can then afterwards check to ensure that os.path.isdir() to ensure that the result really is a directory.